### ISSUE TITLE
Change pan trap species identification form

### ISSUE THREAD

**Issue by JordyvdB97** [BiologicalRecordsCentre/SPRING#95]
Hi everyone, 

Last week David asked me to explain and raise this issue myself here on GitHub. 

When entering the "Select pan-trap samples to add insect SPECIES data" I can add my identifications, including the specimen code and sex. However when I send my materials to a colleague for verification and he/she needs to change only one identification it is very troublesome to look up the correct pan trap in the list, than the correct specimen, and than change the identification. For the next specimen someone needs to search for another pan trap in the list ... and so on. 

(1) Can we add a table/form where you can simply search for a specimen code and than change the identification (and maybe "Sex", "Identified By", and "Comments"). And this way changes the "Select pan-trap samples to add insect SPECIES data"-form of that specimen accordingly. 

In our institute we use specimen code labels with a QR-code. So this speeds up the identification verification process for our subcontractors a lot. Probably for other institutions as well. 

In the slipstream of this I have a few other suggestions:

(2) Can we add all bee / hoverfly genera, families, and bee unknown to the species list? Than we have a placeholder for specimens that await further identification. 

(3) Specimen code only allows whole numbers and should allow for most institutions also letters and special characters. Our specimen codes are like: RMNH.INS.943304.

(4) "Sex" should include "worker" and "queen" for social bees. 

(5) Good to have a field next to "Identified By" called "Verified By". Both fields should include the year of identification / verification (either as separate field or it should be entered every time in the name field).

Thanks a lot in advance. 

Jordy van der Beek

**Comment by andrewvanbreda** [comment]
Hi @JordyvdB97
In an attempt to make your life a bit easier as quickly as I can, the Specimen Code has been added as a column to both the Bycatch Data and Species Data page grids.
This is searchable by filling in the filter box directly under the column header, and then pressing return.

**Comment by andrewvanbreda** [comment]
Sorry @JordyvdB97,
Ignore above comments (which I will now delete), I was getting confused with the sample code field.

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod,

Could you run these 4 SQL statements for me please whenever you get a chance.

Some attribute changes have been requested, the problem is in both cases we are already relying on a general Indicia shared attribute, so need to determine to what extent the current attributes are in use before possibly creating new one, and copy the existing data across.

Cheers. I hope the sql is right first time, but we'll see.

select count(*) 
from indicia.occurrence_attribute_values oav
join indicia.occurrences o on o.id = oav.occurrence_id and o.deleted=false
join indicia.samples s on s.id = o.sample_id AND s.survey_id = 670
and s.deleted=false
where oav.occurrence_attribute_id = 712
AND oav.deleted=false;
**Result : 2**

select distinct s.id
from indicia.occurrence_attribute_values oav
join indicia.occurrences o on o.id = oav.occurrence_id and o.deleted=false
join indicia.samples s on s.id = o.sample_id AND s.survey_id = 670
and s.deleted=false
where oav.occurrence_attribute_id = 712
AND oav.deleted=false
limit 10;
**Result :  20473333**

select count(*) 
from indicia.occurrence_attribute_values oav
join indicia.occurrences o on o.id = oav.occurrence_id and o.deleted=false
join indicia.samples s on s.id = o.sample_id AND s.survey_id = 670
and s.deleted=false
where oav.occurrence_attribute_id = 105
AND oav.deleted=false;
**Result : 8**

select distinct s.id
from indicia.occurrence_attribute_values oav
join indicia.occurrences o on o.id = oav.occurrence_id and o.deleted=false
join indicia.samples s on s.id = o.sample_id AND s.survey_id = 670
and s.deleted=false
where oav.occurrence_attribute_id = 105
AND oav.deleted=false
limit 10;
**Results : 
19993923
20473333**

**Comment by andrewvanbreda** [comment]
@JordyvdB97 @DavidRoy 
I have created a new Specimen Code field which will accept any text.
This is now live.
I have had to leave the old field on the page for now which has been renamed as "Specimen Code (legacy)".
I have put a note on the page to explain that the new field should be used, but the old field will be present until we can copy the data in the database across to the new field.
(I have also noted that it doesn't matter if the old field is used by accident, as we will be copying the data across)

**Comment by BirenRathod** [comment]
@andrewvanbreda See above my comments.

**Comment by andrewvanbreda** [comment]
@BirenRathod Thanks Biren, on intial reading these seem like exactly the kind of results we want. 
I misread the result for a second and thought there was 19993923 records!

**Comment by andrewvanbreda** [comment]
Point 3 is now complete.
The existing values have been copied over (they only seemed to be test values anyway).
I have removed the old version of the field and also removed the notes related to that.

**Comment by andrewvanbreda** [comment]
@DavidRoy @JordyvdB97 
I have also now deployed a new Sex field for bees.
It will always include the extra options because I do not think I can determine what a social bee would be from the species list which has 2000+ entries.
Unless you have any idears how I would determine that logically we can leave it down to the user making the decision, it probably isn't worth the programming budget to make the termlist context sensitive.

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod 
Could you rerun the 4 SQL statements Biren so I can see there isn't any old data to copy across to the new fields thanks. Cheers (not urgent)

**Comment by BirenRathod** [comment]
@andrewvanbreda Below are the results.

Results :
Query 1: 0

Query 2: No result

Query 3: 5

Query 4:
19993923
20473333

**Comment by andrewvanbreda** [comment]
@BirenRathod 
Thanks Biren, that result is good, as the 5 rows there relate to the Hoverfly grid which still uses the original Sex termlist.
Thanks for your help.

**Comment by andrewvanbreda** [comment]
Hi @JordyvdB97 @DavidRoy 
David: Note Jordy's request:

"(2) Can we add all bee / hoverfly genera, families, and bee unknown to the species list? Than we have a placeholder for specimens that await further identification."

I have just checked, this species list is the "Bees in Europe" list.
http://warehouse1.indicia.org.uk/index.php/taxon_list/edit/273?tab=taxa

This looks like a controlled list to me (i.e. not one I should just edit).
I don't edit lists unless they are clearly OK for me to edit (such as one I created myself for a specific project).
Are you able to consider Jordy's request and take any action you see fit, taking into account this list may be used by several projects (I am not sure)

**Comment by DavidRoy** [comment]
@andrewvanbreda - please go ahead and extend the Bee and Hoverfly lists as requested by Jordy.  These lists are not used by other projects

**Comment by andrewvanbreda** [comment]
@DavidRoy @JordyvdB97 
Hi David, thanks for the feedback. 
Am a little confused what entries are needed. 
I will create a "Bee unknown" entry right now.
I am unsure of what to create for "all bee / hoverfly genera, families".

**Comment by DavidRoy** [comment]
@andrewvanbreda the higher taxa for bees and hoverflies need to be available as part of the 'species' list for recording.  Some data cannot be identified to species

**Comment by andrewvanbreda** [comment]
@DavidRoy @JordyvdB97
Ok, I have added a new entry to each species list
Hoverfly (not identified)
Bee (not identified)

**Comment by JordyvdB97** [comment]
@DavidRoy @andrewvanbreda 
All the changes so far work very well. Thanks a lot! 
In addition to the Hoverfly (not identified) and Bee (not identified), it is also good to have a list of all genera (and for bees families as well). I made a two lists with all genera that are not-monotypic, which can be used to add those as new entries. 

[Bee_genera_families.txt](https://github.com/BiologicalRecordsCentre/SPRING/files/9926987/Bee_genera_families.txt)
[hoverfly_genera.txt](https://github.com/BiologicalRecordsCentre/SPRING/files/9926988/hoverfly_genera.txt)

In addition to that, I need two new entries for the specieslist (aggregated species)

For Bees:
Bombus terrestris/cryptarum/lucorum/magnus/

For Hoverflies:
Eumerus sogdianus/strigatus

**Comment by DavidRoy** [comment]
Thanks for testing @JordyvdB97 
@andrewvanbreda can you make these further additions to the species lists please

**Comment by andrewvanbreda** [comment]
Hi @DavidRoy @JordyvdB97
Those two aggregated species added.
The files have been imported and are now live.

**Comment by andrewvanbreda** [comment]
Note the verification information addition is coded an committed, but unable to put live until the Warehouse database is updated.

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod 
Would you be able to pull these two reports whenever you get a chance please. Thanks.

reports/projects/PoMS/select_occs_to_add_insect_data_by_specimen_code.xml
reports/projects/PoMS/select_occs_to_add_insect_data_by_specimen_code_mapping.xml

**Comment by BirenRathod** [comment]
@andrewvanbreda Those reports gone already on live.

**Comment by andrewvanbreda** [comment]
@BirenRathod 
Thanks Biren :) Oh that was quick then if  already there, has had only just pushed them. Maybe someone pulled for another project.

**Comment by BirenRathod** [comment]
the usaul suspect, @johnvanbreda.

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod,

Could you pull the reports again whenever you get a chance please,

reports/projects/PoMS/select_occs_to_add_insect_data_by_specimen_code.xml
reports/projects/PoMS/select_occs_to_add_insect_data_by_specimen_code_mapping.xml

Thanks

**Comment by BirenRathod** [comment]
@andrewvanbreda They are gone now on live.

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod 
Have made a couple of adjustments. Would you be able to pull those reports again please whenever you get a sec.
Thanks :)

**Comment by BirenRathod** [comment]
@andrewvanbreda they have been updated now.

**Comment by andrewvanbreda** [comment]
Hi @JordyvdB97
On the "Select pan-trap samples to add insect SPECIES data" search page
I have put a new link which will take you to a page that allows you to search by specimen code.
https://pollinator-monitoring.net/select-pan-trap-samples-add-insect-species-data
The link is at the top of the page says
"If you would prefer to search by existing specimens, click here""
Note that you do still have to click on the edit icon to enter the data entry page for now, but perhaps sometime in future we can look at allowing some of the data to be edited directly on the grid, but that is unlikely to happen in the very short term.

**Comment by JordyvdB97** [comment]
Hi @andrewvanbreda,

Sorry for the long silence. But I just checked the new search page with a few specimens and it works very smoothly. 
Thank you very much for organising. 

@DavidRoy There is only one request from my side left. Do you think it is a good idea if we can **have a field next to "Identified By" called "Verified By"**. Good to track who have seen the specimens in the system. Another good **addition will be the year (or date) of identification and verification**. When species will be split, or lumped, we can easily see which specimens need to be reexamined. However, this way we might end up with too much columns in the end. What do you think?

Kind regards, Jordy

**Comment by andrewvanbreda** [comment]
Hi @JordyvdB97 @DavidRoy,

I have actually been wanting to get back to you about the verifier field.
I actually did some coding for this and completed that, but then I realised a possible problem during testing, so I didn't put it on the page.
I spoke to Martin Harvey who runs the UKPoMS system and he confirmed my thoughts.
That system uses very similar pages for adding bycatch and adding expert species data.

When an expert adds their entries onto that species data page, those occurrences don't actually enter the iRecord verification system.
So actually, if I add a verified By column to that grid, it won't actually ever contain anything.
The Identified By column is the name of the expert adding the species data, and then it doesn't actually go any further than that.
Perhaps this column needs renaming if it is confusing people?

I think the only person not displayed on that grid at the moment is the original Recorder Name, but that is at the top of the form, and so is the sampling date.

There is one thing you mentioned I haven't talked about above, and that is the identification date, but I will address that when you have taken a look at what I have just said.

Anyway let me know your thoughts on what I have just said.

Andy
